libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala

```diff
@@ -38,11 +38,12 @@ object DataType {
       ValidCommit()
     )
 
-  final case class ValidEcuSerial()
-  type EcuSerial = Refined[String, ValidEcuSerial]
-
-  implicit val validEcuSerial: Validate.Plain[String, ValidEcuSerial] =
-    validInBetween(min = 1, max = 64, ValidEcuSerial())
+  final case class EcuIdentifier(value: String) extends AnyVal {
+    def apply(v: String): EcuIdentifier = {
+      require(value.length >= 1 && value.length <= 64, s"$value is not between 1 and 64 chars long.")
```

Comment:
Result of validation should be `Either[ValidationError, EcuIdentifier]`, require throws an exception

Reply:
I suggested we use require, but Either is a better option.

The thing about using `object.apply` is that IIRC our codecs will use shapeless which bypasses `apply` and this validation...

Reply:
require throws when you try to create the object so the first line throws.

Reply:
What I am thinking of is having something like
```
trait SmartConstuctor[A, B] {
  def apply(a: A): Either[ValidationError, B]
}
```
and try to derive circe, slick and akka-http from it

Reply:
@simao it does not. Apply does not have any special semantic in a class definition

Reply:
Sorry, I actually don't understand the question.

`ecu` is not a function so...

Reply:
@koshelev wrt the SmartConstructor thing, yeah that makes sense, I was just mentioning we cannot just use the derivation we currently have, using shapeless. Though I think that they might actually work if you make the constructor of the case class private, shapeless will use `apply` from the companion object.

Reply:
@simao an `apply(...)` method is a syntactic sugar in scala. `ecu("invalid value")` is the same as `ecu.apply("invalid value")`

Reply:
I know.. but
```
val ecu: EcuIdentifier = EcuIdentifier("valid input")
ecu("invalid value") // throws exception
```

here `ecu` is an instance of EcuIdentifier, it doesn't have an apply method... Unless I am reading this all wrong.

Reply:
Are not we discussing https://github.com/advancedtelematic/libats/pull/82#discussion-diff-239468700R41?

Reply:
No worries. Should we put this topic aside for some time and do more thinking on it?

Reply:
Sure, I implemented something in the delegations PR but I don't think it's optimal, needs to be improved.